### ISSUE TITLE
clean up KNN

### ISSUE THREAD

**Issue by karlnapf** [shogun-toolbox/shogun#3507]
Currently, there are multiple solvers used in KNN.
All of this is done in a flat structure in a single implementation file.
This task is to build a version that respects the object oriented paradigm, i.e. each implementation is done in a different (sub)class. This will increase code readability, cleanliness, and ease of debugging.

This is mostly copy paste plus some minimal standard design decisions

**Comment by karlnapf** [comment]
Another nice task is to clean up all the old code in there (use SGVector rather than pointers), and make use of linalg or eigen3 rather than all these loops.
Finally, replace pthreads with C++11 threads or openmp

**Comment by MikeLing** [comment]
Hey @karlnapf, I would like to work on this issue, is that about to separate https://github.com/shogun-toolbox/shogun/blob/9a73256357e356cc661852874c2c3ff09ffb6d6b/src/shogun/multiclass/KNN.cpp into different files(subclasses)?   Should I still put them into shogun/src/shogun/multiclass or I should  put them into a new path like shogun/src/shogun/KNN folder?

**Comment by MikeLing** [comment]
Oh furthermore, I found PR #3544 is closed rather than merged. So, Do we still need replace pointer and use linalg instead of loops in KNN?

Thank you :)

**Comment by karlnapf** [comment]
Yes that's all still useful to clean up. Start with something simple.... not everything at once

**Comment by MikeLing** [comment]
A small question in here, why we don't use semicolon after `ASSERT()`(e.g. https://github.com/shogun-toolbox/shogun/blob/9a73256357e356cc661852874c2c3ff09ffb6d6b/src/shogun/multiclass/KNN.cpp#L47)

**Comment by MikeLing** [comment]
what's the exactly meaning of **lhs** and **rhs** feature in KNN code?

**Comment by MikeLing** [comment]
@karlnapf could you give me some hits about  lhs and rhs feature in shogun ? Thank you :)

**Comment by karlnapf** [comment]
Let's discuss such questions in IRC and keep the thread clean of them.

**Comment by glitch401** [comment]
Hey @karlnapf , can I solve this problem along with issue #3513
